# COR-IS1702 Computational Thinking (2022/23 Term 1)
# Group Project: Finding a high point on a 2D map
---
## (1) Changes to this document
* Significant changes to this document will be listed here.

## (2) General Instructions
* Requirements (this document) release: 19/Oct/2022 (Week 10, Wed)
* Deadline: **9/Nov/2022** (Week 13, Wed) 23:00 hrs. **Late submissions will be awarded zero.**
* Teams: You would have been assigned to a team of 2 (or 3) by your faculty. 

## (3) Warning

* Cheating, plagiarism and dishonesty will be severely punished and filed officially as a disciplinary record with the University.
* Distribution of solutions is equally culpable as the copying of solutions. Do **not** share solutions/partial solutions with other teams, including posting code at any forum or public code repository (e.g. GitHub with public access).
* When in doubt as to whether an action is considered dishonest, do consult your faculty member.
* You must acknowledge usage of 3rd party code/libraries in your report.

## (4) Deliverablies
* You are required to submit to eLearn a single zip file (`Gx_Ty.zip` - where `x` is your section, and `y` is your team number) that contains: 
>* Your `Gx_Ty.py` file (with the `findHighPoint()` function), and 
>*  a report in the form of a PowerPoint file. You may submit a PDF file or a ppt/pptx file (`Gx_Ty.pptx/pdf`).
* Any member of the team can submit, but only the latest submission (by any member) will be stored and used for grading.
* Submission: to [eLearn (Merged)](https://elearn.smu.edu.sg/d2l/home/334420)->Assignments
* Your faculty may select outstanding teams to present their solution (report) to the class. Please check with your respective faculty concerning this.

## (5) Scoring
* Weight: This project is worth 20% of your course grade. In some cases, individual members of each team may not be awarded the same score.
* Rubric:
>1. Quality score (12%)
>2. Running time (5%). Your code should **not** take more than 1 minute to complete running on colab.
>3. Report (3%)

## (6) Report
* Cover. Indicate group (`Gx_Ty`) and member names on the cover slide.
* Content (max: 3 slides) 
>* Explain your algorithm and how it works (you may use pseudocode)
>* Derive the worse-case Big O time complexity of your algo (show your assumptions and working)
>* Focus on clarity
* Contribution: Include a slide that shows the specific contribution of each member
* References: Include references if relevant. Include URLs if available in your references. You **must** acknowledge usage of 3rd party code/libraries and ideas to avoid plagarism.

## (7) Dealing with team issues
* One of the objectives of group projects is to learn how to work effectively in a team. Do ensure that all members contribute constructively to the team effort.
* If a team member does not contribute or does not respond, please alert your faculty as early as possible so that remediation action can be taken.
* You are required to include a section in your project report that briefly describes individual contributions to the project.
* After the project deadline, selected team members may be interviewed by your respective faculty. Do preserve evidence of communication and contribution until the end of this term. 

## (8) Context

You are given a 2D map (size $\text{W} \times \text{H}$), containing mountains, hills, valleys, rivers, oceans, etc. You are also given a function `getElevation(x, y)` to get the elevation (height) of a given point on the map ($x$, $y$), where $0\leq x < \text{W}$ and $0\leq y < \text{H}$. 

Your task is to design an efficient algorithm to find a high point in the map (the optimal solution is the highest point on the map).

We will count the number of times your algorithm calls `getElevation(x, y)` as a measure of how efficient your algorithm is.

---
## (9) Scoring:

Your algorithm will be evaluated in two ways:
- Running time (5%). The algorithm execution time on colab must **not** exceed 1 minute (0 mark if it exeeds 1 min). On the same map, we will run the algorithm 10 times and take the average execution time for evaluation. You will get a higher score for a shorter running time.

- Quality score (12%). You will get a higher score for a higher quality score. Your quality score is computed as follows:

$$
\text{Quality Score}=\text{Elevation} \times (1 - \text{Punishment Ratio})
$$
where **elevation** is the height of the point returned by your algorithm, 
and:
$$
\text{Punishment Ratio}=
\begin{cases}
0 & \quad \text{number of calls to getElevation(x,y): $count<0.1\% \times W \times H$ }\\ 
10 \times \frac{count}{W \times H} & \quad \text{otherwise}
\end{cases}
$$
<br>

## Scoring Example: 
* For a particular map, W = 100, H = 100. So, 0.1% x W x H = 10
>* If your code calls `getElevation()` fewer than 10 times, 
>>* punishment ratio = 0, and 
>>* Quality score = Elevation x (1)
>* If your code calls `getElevation()` 200 times, 
>>* punishment ratio = 10 x 200/(100x100) = 0.2, and 
>>* Quality score = Elevation x (1 - 0.2)
>* If your code calls `getElevation()` 500 times, 
>>* punishment ratio = 10 x 500/(100x100) = 0.5, and 
>>* Quality score = Elevation x (1 - 0.5)

## (10) Sample Maps:
You are given 5 maps for testing purposes. Each map is basically a text file that contains the heights of every single point on the map. You are only allowed to retrieve the height of a certain point using the `getElevation()` function because the number of calls to `getElevation()` will be tracked in order to compute the punishment ratio. 

After the submission deadline, your submission will be scored using new maps that are not given to you of similiar dimensions. A good algorithm is expected to work well with novel maps.

---
Before you start, run the next code fragment to download 5 CSV (comma separated values) files that you will need to your colab environment, as well as some functions that are required.

In [3]:
from urllib.request import urlretrieve
import os

# represents a map class
class Map:
  def __init__(self, fpath, delay=0.00075):
    self._elevations, self.W, self.H = self._read_map(fpath)
    self._counter = 0
    self.delay = delay

  def reset_counter(self):
    self._counter = 0

  def _read_map(self, fpath):
    elevations = []
    with open(fpath, 'r') as f:
      for l in f:
        tokens = l.strip().split()
        elevations.append([int(tok) for tok in tokens])
    return elevations, len(elevations), len(elevations[0])
  
  # returns the elevation of (x,y)
  def getElevation(self, x, y):
    '''
    x and y should be integer
    '''

    if x < 0 or x >= self.W or y < 0 or y >= self.H:
      self.reset_counter()
      raise Exception("Sorry, coordinate out of bound x={}, y={}. Boundary 0 <= x < {}, 0 <= y < {}".format(x, y, self.W, self.H))
    self._counter += 1
    if self.delay > 0:
      import time
      time.sleep(self.delay)
    return self._elevations[x][y]

  @property
  def counter(self):
    return self._counter

# ------------------------------------------------------------------------------
# check if answer is OK (returns None if OK, else returns an error msg)
def get_error_message(answer):
  if answer == None:
    return "Error : your function returned None. It should return a list/tuple of 2 integers."
  elif type(answer) not in [list, tuple]:
    return "Error : your function returned something other than a list/tuple. It should return a list of 2 integers."
  elif len(answer) != 2:
    return "Error : your function returned a list/tuple of fewer than, or more than 2 elements. It should return a list/tuple of exactly 2 integers."
  elif not all(isinstance(i, int) for i in answer):  # check if all elements in answer are int
    return "Error : your function returned a list/tuple of elements, but not all of them are integers. It should return a list/tuple of exactly 2 integers."
  else:
    return None  # no problem, no error msg

# ------------------------------------------------------------------------------    
# returns quality score of an answer
def compute_quality_score(answer, given_map, threshold=0.001):
  '''
    The quality score is equal to the elevation that the designed algorithm found.
    However, this quality score will be reduced by a punishment ratio.
  '''
  W = given_map.W
  H = given_map.H
  counter = given_map.counter
  elevation = given_map.getElevation(answer[0], answer[1])
  total_points = W * H
  punishment_factor = 0
  if counter / total_points > threshold:
    punishment_factor = counter / total_points
  return max(elevation - elevation * 10 * punishment_factor, 0)

# Download a map
def download(url, file):
    if os.path.isfile(file):
        print(file + " already downloaded. You can see it if you click on the folder icon at the left")
    else:
        print("Downloading file " + file + " ...", end="")
        urlretrieve(url,file)
        print("OK")

# download 5 maps for testing
download('https://github.com/lthoang/CT-T1-AY22-23/raw/main/data/map1.txt', 'map1.txt')
download('https://github.com/lthoang/CT-T1-AY22-23/raw/main/data/map2.txt', 'map2.txt')
download('https://github.com/lthoang/CT-T1-AY22-23/raw/main/data/map3.txt', 'map3.txt')
download('https://github.com/lthoang/CT-T1-AY22-23/raw/main/data/map4.txt', 'map4.txt')
download('https://github.com/lthoang/CT-T1-AY22-23/raw/main/data/map5.txt', 'map5.txt')

map1 = Map('map1.txt')
map2 = Map('map2.txt')
map3 = Map('map3.txt')
map4 = Map('map4.txt')
map5 = Map('map5.txt')
map7 = Map('map7.txt')

map1.txt already downloaded. You can see it if you click on the folder icon at the left
map2.txt already downloaded. You can see it if you click on the folder icon at the left
map3.txt already downloaded. You can see it if you click on the folder icon at the left
map4.txt already downloaded. You can see it if you click on the folder icon at the left
map5.txt already downloaded. You can see it if you click on the folder icon at the left


In [9]:
print(map7.getElevation(0,0))

1000


The following diagrams give a idea of the elevation profile of the 5 maps that you have downloaded:

Map 1.

![map1.png](https://github.com/lthoang/CT-T1-AY22-23/raw/main/data/map1.png)

Map 2.

![map2.png](https://github.com/lthoang/CT-T1-AY22-23/raw/main/data/map2.png)

Map 3.

![map3.png](https://github.com/lthoang/CT-T1-AY22-23/raw/main/data/map3.png)

Map 4.

![map4.png](https://github.com/lthoang/CT-T1-AY22-23/raw/main/data/map4.png)

Map 5.

![map5.png](https://github.com/lthoang/CT-T1-AY22-23/raw/main/data/map5.png)


## (11) Requirement
You are required to implement your algorithm in the following function (`findHighPoint`). This function takes in a map and returns your solution (`x, y`) on the map. 

Remember that your algorithm is **not** expected to return the optimal solution (i.e. the highest point on the map), but a good algorithm will return a near-optimal solution. In other words, it is definitely OK if your algorithm does not result in the "real best solution".

In [16]:
# Algo 2

# Write your algorithm in this function. 
# Except for import statements, all other code must be within the findHighPoint function or additional helper functions
import math

def findHighPoint(given_map):
 
    # Initialize variables
    W = given_map.W
    H = given_map.H
    quality_score = W*H*0.1/100
#     print(W, H)
    
    max_height = [0,0,0]
    top3_cord = [0,0,0]
    x_cord = 0
    y_cord = 0
    count = 0
    
    # Space out the coordinates equally
    for x in range(1, W, math.floor(W/(W+H)* 68)):
        for y in range(1, H, math.ceil(H/(W+H)* 68)):
#             if count >= quality_score:
#                 print("break")
#                 return [x_cord, y_cord]
        
            temp_height = given_map.getElevation(x, y)
            print(x,y,temp_height, max_height)
            count += 1
            
            if temp_height > min(max_height) and temp_height not in max_height:
                max_height[max_height.index(min(max_height))] = temp_height
                top3_cord[max_height.index(min(max_height))] = [x, y]
#                 x_cord = x
#                 y_cord = y
    
    print("top 3 height: " + str(max_height))
    print("top 3 cord: " + str(top3_cord))
    prev_point = [x_cord, y_cord]
    
    temp_max = max_height.copy()
    temp_max.sort(reverse = True)
    for x in temp_max:
        x_cord = top3_cord[max_height.index(x)][0]
        y_cord = top3_cord[max_height.index(x)][1]
        max_height_temp = max_height.index(x)
        print("test: " + str(x_cord) + str(y_cord))
        
        while count != quality_score:
        
            north = [x_cord, min(y_cord + 1, H)]
            east = [min(x_cord + 1, W), y_cord]
            south = [x_cord, max(y_cord - 1, 0)]
            west = [max(x_cord - 1, 0), y_cord]
        
            points = [north, east, south, west]
            if prev_point in points:
                points.remove(prev_point)
        
            count += len(points)
            if count >= quality_score:
                break
            
            elevations = [given_map.getElevation(x[0], x[1]) for x in points]
            elevations.append(max_height_temp)
            points.append([x_cord, y_cord])
            max_point = points[elevations.index(max(elevations))]
        
            if (max_point[0] == x_cord and max_point[1] == y_cord):
                print(f"break - q_score: {quality_score}, count: {count}")
                break
#                 return [x_cord, y_cord]
        
            else:
                prev_point = [x_cord, y_cord]
                x_cord = max_point[0]
                y_cord = max_point[1]
                max_height_temp = max(elevations)
                print(x_cord, y_cord, max_height_temp)
    
#     print(W, H)
    print(f"q_score: {quality_score}, count: {count}")
    return [x_cord, y_cord] # return highest point

For submission, copy the whole chunk of code in the previous code box and paste the text into `TeamName.py` (e.g. `G1_T01.py`). Zip this file together with your report (`TeamName.zip`) and submit a single zip file to eLearn. Only the latest submission by any member of your team will be kept on eLearn.

In [17]:
# Test case
file_name = "map3.txt"  # <---- CHANGE CSV FILE NAME HERE to use a different map

import time, copy
# (1) ----- prepare data ------
print("(1) Reading data from " + file_name + " now...\n")
given_map = Map(file_name)

# (2) ----- run the test case ------
print("(2) Starting timer...")
print("Calling your function now using the map read from " + file_name)
start_time = time.time()
answer = findHighPoint(given_map) # calls your function
time_taken = time.time() - start_time
print("Stopping timer...")
print("Execution time " + str(time_taken) + " seconds.\n")    # display time lapsed

# (3) ----- correctness testing code ------ 
print("(3) Checking your answer...")
error_message = get_error_message(answer)

if error_message == None:
  # all is good
  print("Your function returned a valid answer")
  quality_score = compute_quality_score(answer, given_map)
  print("Returned coordinate:", answer)
  print("Elevation (height):", given_map.getElevation(answer[0], answer[1]))
  print("Quality score:", quality_score)
else:
  # there is an error
  print("Your function is not correctly written :-(")
  print(error_message)

(1) Reading data from map3.txt now...

(2) Starting timer...
Calling your function now using the map read from map3.txt
1 1 859 [0, 0, 0]
1 39 732 [859, 0, 0]
1 77 745 [859, 732, 0]
1 115 643 [859, 732, 745]
1 153 718 [859, 732, 745]
1 191 808 [859, 732, 745]
1 229 701 [859, 808, 745]
1 267 759 [859, 808, 745]
1 305 760 [859, 808, 759]
1 343 810 [859, 808, 760]
1 381 737 [859, 808, 810]
1 419 608 [859, 808, 810]
1 457 558 [859, 808, 810]
1 495 484 [859, 808, 810]
1 533 412 [859, 808, 810]
1 571 351 [859, 808, 810]
1 609 287 [859, 808, 810]
1 647 376 [859, 808, 810]
1 685 478 [859, 808, 810]
1 723 490 [859, 808, 810]
1 761 432 [859, 808, 810]
1 799 291 [859, 808, 810]
1 837 218 [859, 808, 810]
1 875 214 [859, 808, 810]
1 913 230 [859, 808, 810]
1 951 255 [859, 808, 810]
1 989 237 [859, 808, 810]
31 1 814 [859, 808, 810]
31 39 784 [859, 814, 810]
31 77 696 [859, 814, 810]
31 115 616 [859, 814, 810]
31 153 609 [859, 814, 810]
31 191 736 [859, 814, 810]
31 229 657 [859, 814, 810]
31 267 74

451 381 332 [920, 933, 939]
451 419 361 [920, 933, 939]
451 457 454 [920, 933, 939]
451 495 552 [920, 933, 939]
451 533 475 [920, 933, 939]
451 571 507 [920, 933, 939]
451 609 589 [920, 933, 939]
451 647 590 [920, 933, 939]
451 685 521 [920, 933, 939]
451 723 590 [920, 933, 939]
451 761 509 [920, 933, 939]
451 799 490 [920, 933, 939]
451 837 446 [920, 933, 939]
451 875 321 [920, 933, 939]
451 913 267 [920, 933, 939]
451 951 167 [920, 933, 939]
451 989 186 [920, 933, 939]
481 1 599 [920, 933, 939]
481 39 624 [920, 933, 939]
481 77 656 [920, 933, 939]
481 115 639 [920, 933, 939]
481 153 563 [920, 933, 939]
481 191 534 [920, 933, 939]
481 229 479 [920, 933, 939]
481 267 464 [920, 933, 939]
481 305 458 [920, 933, 939]
481 343 414 [920, 933, 939]
481 381 409 [920, 933, 939]
481 419 313 [920, 933, 939]
481 457 403 [920, 933, 939]
481 495 458 [920, 933, 939]
481 533 440 [920, 933, 939]
481 571 403 [920, 933, 939]
481 609 564 [920, 933, 939]
481 647 722 [920, 933, 939]
481 685 681 [920, 933, 9

301 3 947
test: 30139
q_score: 800.0, count: 800
Stopping timer...
Execution time 0.8330230712890625 seconds.

(3) Checking your answer...
Your function returned a valid answer
Returned coordinate: [301, 39]
Elevation (height): 933
Quality score: 933


In [13]:
# Evaluate multiple times (10 times), return the average scores overall.
file_name = "map4.txt"  # <---- CHANGE CSV FILE NAME HERE to use a different map
import time
# (1) ----- prepare data ------
print("(1) Reading data from " + file_name + " now...\n")
given_map = Map(file_name)
# (2) ----- run the test case ------
print("(2) Starting timer...")
print("Calling your function now using the map read from " + file_name)
execute_times = []
quality_scores = []

for _ in range(10):
  given_map.reset_counter()
  start_time = time.time()
  answer = findHighPoint(given_map) # calls your function 
  time_taken = time.time() - start_time
  execute_times.append(time_taken)
  print("Stopping timer...")
  print("Execution time " + str(time_taken) + " seconds.\n")    # display time lapsed

  # (3) ----- correctness testing code ------ 
  print("(3) Checking your answer...")
  error_message = get_error_message(answer)

  if error_message == None:
    # all is good
    print("Your function returned a valid answer")
    quality_score = compute_quality_score(answer, given_map)
    print("Returned coordinate:", answer)
    print("Elevation (height):", given_map.getElevation(answer[0], answer[1]))
    print("Quality score:", quality_score)
    quality_scores.append(quality_score)
  else:
    # there is an error
    print("Your function is not correctly written :-(")
    print(error_message)
    
print('\n')
print('Avg. execution time =', sum(execute_times) / len(execute_times), " seconds")
print('Avg. quality score  =', sum(quality_scores) / len(quality_scores))

(1) Reading data from map4.txt now...

(2) Starting timer...
Calling your function now using the map read from map4.txt
1 1 382 0
1 38 360 382
1 75 329 382
1 112 298 382
1 149 275 382
1 186 321 382
1 223 355 382
1 260 403 382
1 297 375 403
1 334 459 403
1 371 359 459
1 408 509 459
1 445 450 509
1 482 372 509
1 519 311 509
1 556 302 509
1 593 278 509
1 630 332 509
1 667 351 509
1 704 284 509
1 741 232 509
1 778 227 509
1 815 205 509
1 852 288 509
1 889 272 509
1 926 281 509
1 963 249 509
30 1 326 509
30 38 348 509
30 75 325 509
30 112 347 509
30 149 336 509
30 186 271 509
30 223 262 509
30 260 266 509
30 297 304 509
30 334 327 509
30 371 332 509
30 408 354 509
30 445 345 509
30 482 281 509
30 519 239 509
30 556 229 509
30 593 226 509
30 630 246 509
30 667 261 509
30 704 248 509
30 741 241 509
30 778 246 509
30 815 307 509
30 852 374 509
30 889 332 509
30 926 305 509
30 963 323 509
59 1 314 509
59 38 293 509
59 75 284 509
59 112 289 509
59 149 314 509
59 186 270 509
59 223 291 509
59 260

668 334 241 687
668 371 273 687
668 408 299 687
668 445 348 687
668 482 350 687
668 519 312 687
668 556 303 687
668 593 241 687
668 630 230 687
668 667 213 687
668 704 210 687
668 741 211 687
668 778 256 687
668 815 277 687
668 852 322 687
668 889 346 687
668 926 271 687
668 963 286 687
697 1 390 687
697 38 348 687
697 75 418 687
697 112 467 687
697 149 540 687
697 186 526 687
697 223 462 687
697 260 342 687
697 297 242 687
697 334 271 687
697 371 280 687
697 408 331 687
697 445 413 687
697 482 377 687
697 519 367 687
697 556 308 687
697 593 236 687
697 630 247 687
697 667 226 687
697 704 221 687
697 741 219 687
697 778 286 687
697 815 345 687
697 852 307 687
697 889 344 687
697 926 285 687
697 963 286 687
726 1 318 687
726 38 329 687
726 75 364 687
726 112 412 687
726 149 428 687
726 186 393 687
726 223 353 687
726 260 257 687
726 297 347 687
726 334 338 687
726 371 425 687
726 408 373 687
726 445 432 687
726 482 369 687
726 519 285 687
726 556 279 687
726 593 249 687
726 630 279 687


436 593 394 615
436 630 332 615
436 667 295 615
436 704 281 615
436 741 240 615
436 778 211 615
436 815 202 615
436 852 198 615
436 889 213 615
436 926 222 615
436 963 303 615
465 1 667 615
465 38 605 667
465 75 496 667
465 112 429 667
465 149 317 667
465 186 305 667
465 223 328 667
465 260 308 667
465 297 273 667
465 334 284 667
465 371 292 667
465 408 380 667
465 445 406 667
465 482 410 667
465 519 508 667
465 556 523 667
465 593 440 667
465 630 394 667
465 667 350 667
465 704 307 667
465 741 251 667
465 778 205 667
465 815 202 667
465 852 195 667
465 889 254 667
465 926 239 667
465 963 238 667
494 1 672 667
494 38 542 672
494 75 416 672
494 112 361 672
494 149 322 672
494 186 334 672
494 223 293 672
494 260 275 672
494 297 272 672
494 334 297 672
494 371 351 672
494 408 379 672
494 445 378 672
494 482 349 672
494 519 426 672
494 556 462 672
494 593 392 672
494 630 329 672
494 667 299 672
494 704 262 672
494 741 226 672
494 778 205 672
494 815 204 672
494 852 203 672
494 889 279 672


233 149 304 509
233 186 314 509
233 223 302 509
233 260 310 509
233 297 331 509
233 334 297 509
233 371 316 509
233 408 314 509
233 445 355 509
233 482 298 509
233 519 308 509
233 556 364 509
233 593 439 509
233 630 470 509
233 667 416 509
233 704 365 509
233 741 400 509
233 778 308 509
233 815 255 509
233 852 207 509
233 889 207 509
233 926 260 509
233 963 256 509
262 1 333 509
262 38 318 509
262 75 311 509
262 112 335 509
262 149 321 509
262 186 340 509
262 223 335 509
262 260 306 509
262 297 363 509
262 334 356 509
262 371 387 509
262 408 379 509
262 445 407 509
262 482 336 509
262 519 295 509
262 556 342 509
262 593 415 509
262 630 412 509
262 667 432 509
262 704 416 509
262 741 346 509
262 778 306 509
262 815 272 509
262 852 211 509
262 889 199 509
262 926 332 509
262 963 293 509
291 1 336 509
291 38 334 509
291 75 312 509
291 112 339 509
291 149 318 509
291 186 369 509
291 223 371 509
291 260 328 509
291 297 360 509
291 334 399 509
291 371 412 509
291 408 451 509
291 445 419 509


1 630 332 509
1 667 351 509
1 704 284 509
1 741 232 509
1 778 227 509
1 815 205 509
1 852 288 509
1 889 272 509
1 926 281 509
1 963 249 509
30 1 326 509
30 38 348 509
30 75 325 509
30 112 347 509
30 149 336 509
30 186 271 509
30 223 262 509
30 260 266 509
30 297 304 509
30 334 327 509
30 371 332 509
30 408 354 509
30 445 345 509
30 482 281 509
30 519 239 509
30 556 229 509
30 593 226 509
30 630 246 509
30 667 261 509
30 704 248 509
30 741 241 509
30 778 246 509
30 815 307 509
30 852 374 509
30 889 332 509
30 926 305 509
30 963 323 509
59 1 314 509
59 38 293 509
59 75 284 509
59 112 289 509
59 149 314 509
59 186 270 509
59 223 291 509
59 260 247 509
59 297 260 509
59 334 254 509
59 371 274 509
59 408 247 509
59 445 269 509
59 482 236 509
59 519 222 509
59 556 215 509
59 593 217 509
59 630 210 509
59 667 224 509
59 704 215 509
59 741 209 509
59 778 210 509
59 815 278 509
59 852 337 509
59 889 389 509
59 926 365 509
59 963 395 509
88 1 321 509
88 38 302 509
88 75 310 509
88 112 343 509
88

610 482 227 687
610 519 232 687
610 556 237 687
610 593 250 687
610 630 243 687
610 667 221 687
610 704 212 687
610 741 212 687
610 778 207 687
610 815 251 687
610 852 318 687
610 889 369 687
610 926 298 687
610 963 246 687
639 1 408 687
639 38 489 687
639 75 472 687
639 112 615 687
639 149 635 687
639 186 414 687
639 223 361 687
639 260 347 687
639 297 311 687
639 334 248 687
639 371 261 687
639 408 262 687
639 445 243 687
639 482 239 687
639 519 228 687
639 556 230 687
639 593 221 687
639 630 220 687
639 667 215 687
639 704 211 687
639 741 211 687
639 778 225 687
639 815 235 687
639 852 289 687
639 889 396 687
639 926 319 687
639 963 263 687
668 1 409 687
668 38 364 687
668 75 464 687
668 112 497 687
668 149 636 687
668 186 546 687
668 223 352 687
668 260 347 687
668 297 317 687
668 334 241 687
668 371 273 687
668 408 299 687
668 445 348 687
668 482 350 687
668 519 312 687
668 556 303 687
668 593 241 687
668 630 230 687
668 667 213 687
668 704 210 687
668 741 211 687
668 778 256 687


378 889 199 573
378 926 218 573
378 963 301 573
407 1 493 573
407 38 517 573
407 75 477 573
407 112 374 573
407 149 372 573
407 186 317 573
407 223 313 573
407 260 289 573
407 297 281 573
407 334 323 573
407 371 366 573
407 408 380 573
407 445 435 573
407 482 523 573
407 519 494 573
407 556 457 573
407 593 393 573
407 630 309 573
407 667 248 573
407 704 238 573
407 741 251 573
407 778 218 573
407 815 208 573
407 852 198 573
407 889 197 573
407 926 273 573
407 963 228 573
436 1 613 573
436 38 615 613
436 75 545 615
436 112 420 615
436 149 335 615
436 186 315 615
436 223 291 615
436 260 290 615
436 297 295 615
436 334 325 615
436 371 369 615
436 408 395 615
436 445 415 615
436 482 470 615
436 519 608 615
436 556 462 615
436 593 394 615
436 630 332 615
436 667 295 615
436 704 281 615
436 741 240 615
436 778 211 615
436 815 202 615
436 852 198 615
436 889 213 615
436 926 222 615
436 963 303 615
465 1 667 615
465 38 605 667
465 75 496 667
465 112 429 667
465 149 317 667
465 186 305 667
465 

175 408 251 509
175 445 267 509
175 482 302 509
175 519 304 509
175 556 323 509
175 593 377 509
175 630 401 509
175 667 429 509
175 704 329 509
175 741 263 509
175 778 238 509
175 815 225 509
175 852 245 509
175 889 307 509
175 926 241 509
175 963 286 509
204 1 338 509
204 38 314 509
204 75 302 509
204 112 291 509
204 149 310 509
204 186 334 509
204 223 290 509
204 260 295 509
204 297 298 509
204 334 275 509
204 371 264 509
204 408 269 509
204 445 294 509
204 482 274 509
204 519 351 509
204 556 380 509
204 593 464 509
204 630 442 509
204 667 433 509
204 704 310 509
204 741 330 509
204 778 285 509
204 815 238 509
204 852 206 509
204 889 198 509
204 926 198 509
204 963 196 509
233 1 337 509
233 38 317 509
233 75 307 509
233 112 302 509
233 149 304 509
233 186 314 509
233 223 302 509
233 260 310 509
233 297 331 509
233 334 297 509
233 371 316 509
233 408 314 509
233 445 355 509
233 482 298 509
233 519 308 509
233 556 364 509
233 593 439 509
233 630 470 509
233 667 416 509
233 704 365 509


784 186 338 687
784 223 364 687
784 260 350 687
784 297 317 687
784 334 262 687
784 371 415 687
784 408 404 687
784 445 379 687
784 482 344 687
784 519 313 687
784 556 284 687
784 593 246 687
784 630 297 687
784 667 325 687
784 704 328 687
784 741 346 687
784 778 341 687
784 815 294 687
784 852 322 687
784 889 361 687
784 926 489 687
784 963 495 687
577 34 688 687
577 35 691 688
577 36 694 691
577 37 696 694
578 37 695 696
579 37 693 696
580 37 690 696
800 1000
q_score: 800.0, count: 763
Stopping timer...
Execution time 0.8192257881164551 seconds.

(3) Checking your answer...
Your function returned a valid answer
Returned coordinate: [577, 37]
Elevation (height): 696
Quality score: 696
1 1 382 0
1 38 360 382
1 75 329 382
1 112 298 382
1 149 275 382
1 186 321 382
1 223 355 382
1 260 403 382
1 297 375 403
1 334 459 403
1 371 359 459
1 408 509 459
1 445 450 509
1 482 372 509
1 519 311 509
1 556 302 509
1 593 278 509
1 630 332 509
1 667 351 509
1 704 284 509
1 741 232 509
1 778 227 509
1 8

552 741 218 682
552 778 202 682
552 815 207 682
552 852 238 682
552 889 342 682
552 926 297 682
552 963 232 682
581 1 582 682
581 38 687 682
581 75 667 687
581 112 575 687
581 149 454 687
581 186 380 687
581 223 293 687
581 260 281 687
581 297 283 687
581 334 284 687
581 371 243 687
581 408 233 687
581 445 269 687
581 482 243 687
581 519 262 687
581 556 277 687
581 593 314 687
581 630 302 687
581 667 252 687
581 704 244 687
581 741 214 687
581 778 208 687
581 815 218 687
581 852 261 687
581 889 356 687
581 926 318 687
581 963 244 687
610 1 478 687
610 38 618 687
610 75 502 687
610 112 646 687
610 149 527 687
610 186 362 687
610 223 339 687
610 260 373 687
610 297 301 687
610 334 281 687
610 371 237 687
610 408 232 687
610 445 230 687
610 482 227 687
610 519 232 687
610 556 237 687
610 593 250 687
610 630 243 687
610 667 221 687
610 704 212 687
610 741 212 687
610 778 207 687
610 815 251 687
610 852 318 687
610 889 369 687
610 926 298 687
610 963 246 687
639 1 408 687
639 38 489 687
639

349 334 321 509
349 371 339 509
349 408 387 509
349 445 472 509
349 482 517 509
349 519 442 517
349 556 351 517
349 593 282 517
349 630 257 517
349 667 300 517
349 704 357 517
349 741 289 517
349 778 224 517
349 815 196 517
349 852 196 517
349 889 209 517
349 926 304 517
349 963 315 517
378 1 439 517
378 38 424 517
378 75 412 517
378 112 335 517
378 149 326 517
378 186 324 517
378 223 306 517
378 260 298 517
378 297 318 517
378 334 307 517
378 371 301 517
378 408 342 517
378 445 428 517
378 482 573 517
378 519 494 573
378 556 358 573
378 593 368 573
378 630 300 573
378 667 273 573
378 704 322 573
378 741 298 573
378 778 220 573
378 815 192 573
378 852 197 573
378 889 199 573
378 926 218 573
378 963 301 573
407 1 493 573
407 38 517 573
407 75 477 573
407 112 374 573
407 149 372 573
407 186 317 573
407 223 313 573
407 260 289 573
407 297 281 573
407 334 323 573
407 371 366 573
407 408 380 573
407 445 435 573
407 482 523 573
407 519 494 573
407 556 457 573
407 593 393 573
407 630 309 573


117 889 436 509
117 926 363 509
117 963 318 509
146 1 342 509
146 38 322 509
146 75 311 509
146 112 299 509
146 149 292 509
146 186 302 509
146 223 334 509
146 260 341 509
146 297 265 509
146 334 257 509
146 371 242 509
146 408 253 509
146 445 282 509
146 482 271 509
146 519 265 509
146 556 288 509
146 593 285 509
146 630 407 509
146 667 431 509
146 704 322 509
146 741 306 509
146 778 300 509
146 815 282 509
146 852 259 509
146 889 401 509
146 926 324 509
146 963 282 509
175 1 341 509
175 38 322 509
175 75 311 509
175 112 300 509
175 149 290 509
175 186 306 509
175 223 341 509
175 260 293 509
175 297 302 509
175 334 252 509
175 371 251 509
175 408 251 509
175 445 267 509
175 482 302 509
175 519 304 509
175 556 323 509
175 593 377 509
175 630 401 509
175 667 429 509
175 704 329 509
175 741 263 509
175 778 238 509
175 815 225 509
175 852 245 509
175 889 307 509
175 926 241 509
175 963 286 509
204 1 338 509
204 38 314 509
204 75 302 509
204 112 291 509
204 149 310 509
204 186 334 509
204 

726 593 249 687
726 630 279 687
726 667 260 687
726 704 231 687
726 741 263 687
726 778 338 687
726 815 387 687
726 852 322 687
726 889 382 687
726 926 316 687
726 963 356 687
755 1 341 687
755 38 314 687
755 75 307 687
755 112 350 687
755 149 375 687
755 186 392 687
755 223 348 687
755 260 257 687
755 297 261 687
755 334 322 687
755 371 504 687
755 408 475 687
755 445 414 687
755 482 345 687
755 519 305 687
755 556 310 687
755 593 246 687
755 630 301 687
755 667 294 687
755 704 269 687
755 741 321 687
755 778 365 687
755 815 331 687
755 852 311 687
755 889 411 687
755 926 379 687
755 963 446 687
784 1 362 687
784 38 335 687
784 75 324 687
784 112 303 687
784 149 338 687
784 186 338 687
784 223 364 687
784 260 350 687
784 297 317 687
784 334 262 687
784 371 415 687
784 408 404 687
784 445 379 687
784 482 344 687
784 519 313 687
784 556 284 687
784 593 246 687
784 630 297 687
784 667 325 687
784 704 328 687
784 741 346 687
784 778 341 687
784 815 294 687
784 852 322 687
784 889 361 687


523 149 434 672
523 186 383 672
523 223 308 672
523 260 272 672
523 297 265 672
523 334 276 672
523 371 324 672
523 408 291 672
523 445 310 672
523 482 288 672
523 519 378 672
523 556 373 672
523 593 383 672
523 630 329 672
523 667 260 672
523 704 231 672
523 741 220 672
523 778 207 672
523 815 198 672
523 852 206 672
523 889 299 672
523 926 298 672
523 963 249 672
552 1 682 672
552 38 581 682
552 75 524 682
552 112 455 682
552 149 511 682
552 186 437 682
552 223 378 682
552 260 286 682
552 297 263 682
552 334 265 682
552 371 254 682
552 408 248 682
552 445 260 682
552 482 292 682
552 519 306 682
552 556 312 682
552 593 366 682
552 630 331 682
552 667 290 682
552 704 236 682
552 741 218 682
552 778 202 682
552 815 207 682
552 852 238 682
552 889 342 682
552 926 297 682
552 963 232 682
581 1 582 682
581 38 687 682
581 75 667 687
581 112 575 687
581 149 454 687
581 186 380 687
581 223 293 687
581 260 281 687
581 297 283 687
581 334 284 687
581 371 243 687
581 408 233 687
581 445 269 687


<h1><b>Fixed Range</b></h1>

Map 1:
Execution time 12.602922201156616 seconds.
Returned coordinate: [289, 1]
Elevation (height): 532
Quality score: 532

Map 2:
Execution time 12.597789525985718 seconds.
Returned coordinate: [1, 257]
Elevation (height): 793
Quality score: 793

Map 3:
Execution time 12.57199215888977 seconds.
Returned coordinate: [257, 33]
Elevation (height): 932
Quality score: 932

Map 4:
Execution time 12.59573745727539 seconds.
Returned coordinate: [481, 1]
Elevation (height): 718
**Quality score: 718**

Map 5:
Execution time 12.54642915725708 seconds.
Returned coordinate: [129, 1]
Elevation (height): 460
Quality score: 460

<h1><b>Proportion Only</b></h1>

Map 1:
Execution time 12.492062091827393 seconds.
Returned coordinate: [301, 1]
Elevation (height): 549
Quality score: 549

Map 2:
Execution time 12.593671321868896 seconds.
Returned coordinate: [1, 257]
Elevation (height): 793
Quality score: 793

Map 3:
Execution time 12.608817338943481 seconds.
Returned coordinate: [301, 33]
Elevation (height): 950
Quality score: 950

Map 4:
Execution time 12.585641860961914 seconds.
Returned coordinate: [476, 1]
Elevation (height): 717
Quality score: 717

Map 5:
Execution time 12.582794666290283 seconds.
Returned coordinate: [126, 1]
Elevation (height): 467
Quality score: 467

<h1><b>Proportion + Search around highest point</b></h1>

Map 1:
Execution time 0.8448660373687744 seconds.
Returned coordinate: [316, 0]
Elevation (height): 549
Quality score: 549

Map 2:
Execution time 0.8107788562774658 seconds.
Returned coordinate: [0, 259]
Elevation (height): 799 -> 801
Quality score: 801

Map 3:
Execution time 0.8127596378326416 seconds.
Returned coordinate: [290, 7]
Elevation (height): 936
Quality score: 936

Map 4:
Execution time 0.8122210502624512 seconds.
Returned coordinate: [577, 37]
Elevation (height): 696
Quality score: 696

Map 5:
Execution time 0.8143320083618164 seconds.
Returned coordinate: [606, 110]
Elevation (height): 473
**Quality score: 473**

To Improve:
Widen search area for search around highest point

<h1><b>Proportion + Navigator</b></h1>

Map 1:
Execution time 0.8394949436187744 seconds.
Returned coordinate: [310, 1]
Elevation (height): 561
**Quality score: 561**

Map 2:
Execution time 0.8378970623016357 seconds.
Returned coordinate: [0, 263]
Elevation (height): 803
**Quality score: 803**

Map 3:
Execution time 0.842041015625 seconds.
Returned coordinate: [296, 6]
Elevation (height): 962
**Quality score: 962**

Map 4:
Execution time 0.8210921287536621 seconds.
Returned coordinate: [575, 42]
Elevation (height): 702
**Quality score: 702**

Map 5:
Execution time 0.8467440605163574 seconds.
Returned coordinate: [608, 112]
Elevation (height): 470
Quality score: 470

In [22]:
#Map 1: 568 / 568
#Map 2: 803 / 803
#Map 3: 969 / 970
#Map 4: 722 / 722
#Map 5: 468 / 476

#TestMap1: 568 / 704
#TestMap2: 803 / 803
#TestMap3: 970 / 970
#TestMap4: 722 / 772
#TestMap5: 969 / 970